### DISCUSSION TITLE
Docker image created as non root, unable to access files when using docker compose with Error: EACCES: permission denied, open '/opt/key.pem'\n

### DISCUSSION THREAD

**Discussion by nellyk** [community/community#26894] [Actions]
<p>I have this docker file</p>
<pre><code class="lang-auto"># Use the pre-baked fat node image only in the builder
# which includes build utils preinstalled (e.g. gcc, make, etc).
# This will result in faster and reliable One App docker image
# builds as we do not have to run apk installs for alpine.
FROM node:12 as builder
WORKDIR /opt/build
RUN npm install -g npm@6.12.1 --registry=https://registry.npmjs.org
COPY --chown=node:node ./ /opt/build
# npm ci does not run postinstall with root account
RUN NODE_ENV=development npm ci --build-from-source
# npm ci does not run postinstall with root account
# which is why there is a dev build
RUN NODE_ENV=development npm run build &amp;&amp; \
    mkdir -p /opt/one-app/development &amp;&amp; \
    chown node:node /opt/one-app/development &amp;&amp; \
    cp -r /opt/build/. /opt/one-app/development
# prod build
RUN NODE_ENV=production npm run build &amp;&amp; \
    NODE_ENV=production npm prune &amp;&amp; \
    mkdir -p /opt/one-app/production &amp;&amp; \
    chown node:node /opt/one-app/production &amp;&amp; \
    mv /opt/build/LICENSE.txt /opt/one-app/production &amp;&amp; \
    mv /opt/build/node_modules /opt/one-app/production &amp;&amp; \
    mv /opt/build/package.json /opt/one-app/production &amp;&amp; \
    mv /opt/build/lib /opt/one-app/production &amp;&amp; \
    mv /opt/build/build /opt/one-app/production &amp;&amp; \
    mv /opt/build/bundle.integrity.manifest.json /opt/one-app/production &amp;&amp; \
    mv /opt/build/.build-meta.json /opt/one-app/production

# development image
# docker build . --target=development
FROM node:12-alpine as development
ENV NODE_ENV=development
# exposing these ports as they are default for all the local dev servers
# see src/server/config/env/runtime.js
EXPOSE 3000
EXPOSE 3001
EXPOSE 3002
EXPOSE 3005
WORKDIR /opt/one-app
RUN chown node:node /opt/one-app
CMD ["node", "lib/server"]
COPY --from=builder --chown=node:node /opt/one-app/development ./

# production image
# last so that it's the default image artifact
FROM node:12-alpine as production
ENV NODE_ENV=production
# exposing these ports as they are defaults for one app and the prom metrics server
# see src/server/config/env/runtime.js
EXPOSE 3000
EXPOSE 3005
WORKDIR /opt/one-app
CMD ["node", "lib/server"]
COPY --from=builder --chown=node:node /opt/one-app/production ./
</code></pre>
<p>and this is the docker-compose.yml file</p>
<pre><code class="lang-auto">version: '3'

networks:
  one-app-at-test-network:
services:
  one-app:
    build:
      context: ../
      args:
        - http_proxy
        - https_proxy
        - no_proxy
    # tags the built image as:
    image: one-app:at-test
    expose:
      - "8443"
    volumes:
      - ./one-app/one-app-cert.pem:/opt/cert.pem
      - ./one-app/one-app-privkey.pem:/opt/key.pem
      - ./nginx/nginx-cert.pem:/opt/nginx-cert.pem
      - ./extra-certs.pem:/opt/extra-certs.pem
    env_file:
      - ./one-app/base.env
    networks:
      one-app-at-test-network:
    depends_on:
      - "fast-api"
      - "slow-api"
      - "extra-slow-api"
      - "nginx"
    entrypoint: sh -c 'sleep 2s &amp;&amp; node lib/server'
  fast-api:
    build:
      context: ./api
      args:
        - http_proxy
        - https_proxy
        - no_proxy
    ports:
      - "8000:80"
    networks:
      one-app-at-test-network:
        aliases:
          - fast.api.frank
  slow-api:
    build:
      context: ./api
      args:
        - http_proxy
        - https_proxy
        - no_proxy
    ports:
      - "8001:80"
    entrypoint:
      - "npm"
      - "start"
      - "--"
      - "3000"
    networks:
      one-app-at-test-network:
        aliases:
          - slow.api.frank
  extra-slow-api:
    build:
      context: ./api
      args:
        - http_proxy
        - https_proxy
        - no_proxy
    ports:
      - "8002:80"
    entrypoint:
      - "npm"
      - "start"
      - "--"
      - "8000"
    networks:
      one-app-at-test-network:
        aliases:
          - extra-slow.api.frank
  nginx:
    image: nginx:1.17.5-alpine
    volumes:
      - ./nginx/default.conf:/etc/nginx/conf.d/default.conf
      - ./nginx/nginx-cert.pem:/etc/ssl/nginx-cert.pem
      - ./nginx/nginx-privkey.pem:/etc/ssl/nginx-privkey.pem
      - ./nginx/origin-statics:/usr/share/nginx/html
    networks:
      one-app-at-test-network:
        aliases:
          - sample-cdn.frank
  selenium-chrome:
    # specify docker image sha to ensure consistency
    image: selenium/standalone-chrome-debug@sha256:e8bf805eca673e6788fb50249b105be860d991ee0fa3696422b4cb92acb5c07a
    # https://github.com/SeleniumHQ/docker-selenium#running-the-images
    volumes:
      - /dev/shm:/dev/shm
    ports:
      - "4444:4444"
      - "5901:5900"
    networks:
      one-app-at-test-network:
    # sleep 5s to make sure set up is completed prior to opening up chrome
    entrypoint: bash -c '/opt/bin/entry_point.sh &amp; sleep 5s &amp;&amp; google-chrome --ignore-certificate-errors --no-first-run --autofill-server-url https://one-app:8443/success'
</code></pre>
<p>When i start a github action using docker compose i get the following error</p>
<pre><code class="lang-auto">Error: EACCES: permission denied, open '/opt/key.pem'\n
</code></pre>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/c/c32326a74eed7023a48983ce2e72793d92c05210.png" data-download-href="/uploads/short-url/rQgpSr3Ka8rDY47x3vbtYNDG0LK.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/3062772/181100490-d51a1dbe-5d68-4e79-92c1-a4de70187906.png" alt="image" data-base62-sha1="rQgpSr3Ka8rDY47x3vbtYNDG0LK" width="690" height="73" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/c/c32326a74eed7023a48983ce2e72793d92c05210_2_690x73.png, https://aws1.discourse-cdn.com/github/original/2X/c/c32326a74eed7023a48983ce2e72793d92c05210.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/c/c32326a74eed7023a48983ce2e72793d92c05210.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/c/c32326a74eed7023a48983ce2e72793d92c05210_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1010×108 41.2 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>If i remove the user option on the docker file and allow it to be built with root as the user the actions works. How can i keep the USER: node and still be able to access the above files</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="nellyk" data-post="1" data-topic="129404">
<div class="title">
<div class="quote-controls"></div>
 nellyk:</div>
<blockquote>
<p>How can i keep the USER: node and still be able to access the above files</p>
</blockquote>
</aside>
<p>Adjust the access rights on the file so the <code>node</code> user can read it. A few notes on that:</p>
<ul>
<li>You’ll have to do the adjustment outside the container.</li>
<li>The UID the <code>node</code> user has is defined <em>inside</em> the container. It might belong to another username or none at all outside.</li>
<li>That file seems to be a cryptographic key, so DO NOT make it globally readable. The easiest thing is probably to adjust the file owner, but that’s a  security risk if the UID is in use outside the container for unrelated purposes.</li>
</ul>

**Comment by nellyk** [comment]
<aside class="quote no-group" data-username="airtower-luna" data-post="2" data-topic="129404">
<div class="title">
<div class="quote-controls"></div>
 airtower-luna:</div>
<blockquote>
<ul>
<li>That file seems to be a cryptographic key, so DO NOT make it globally readable. The easiest thing is probably to adjust the file owner, but that’s a security risk if the UID is in use outside the container for unrelated purposes.</li>
</ul>
</blockquote>
</aside>
<p>Thank you for responding and the suggestions <a class="mention" href="/airtower-luna">@airtower-luna</a> might you have a suggestion on how i can achieve this</p>
<blockquote>
<p>You’ll have to do the adjustment outside the container.</p>
</blockquote>

**Comment by airtower-luna** [comment]
<p>In principle:</p>
<ol>
<li>Get the UID of the <code>node</code> user inside the container:</li>
</ol>
<pre><code class="lang-auto">docker exec -ti --user node CONTAINER_ID id
</code></pre>
<ol start="2">
<li>Use <code>chown</code> to make the <code>one-app/one-app-privkey.pem</code> file owned by that user.</li>
</ol>
<p>You can use a similar approach to use group access rights instead (with <code>chgrp</code> and possibly <code>chmod</code> to enable read rights for the user.</p>
<p>But: If some other user on the system uses the same UID (or GID in case of group access) they will gain access to the file, which you probably don’t want. Alternatively you could also change the UID of the <code>node</code> user in the container to that if the current file owner, but that might create the reverse problem and let the container read more files than it should.</p>
<p>Also keep in mind that none of these solutions is portable, in case you want to run the same container image on another host. For that it’d be better to <a href="https://docs.docker.com/storage/volumes/">use a named volume</a> to store the certificate and key, so you can adjust access rights in the container scope only.</p>

**Comment by nellyk** [comment, marked as answer]
<p>Thank you so much for the  suggestions.</p>
<p>I eventually went ahead to use ARG and ENV variables and passing them to the docker image depending on what i was using it for<br>
User root for integrations tests<br>
User node for building the image to be distributed</p>
